### PR TITLE
dev_mode: Reliable plugin reloads

### PR DESCRIPTION
**Pull request by abusalimov**
This PR implements reloading modules in a more robust and reliable way using import hooks. The key feature is that the modules get reloaded in the very same order they would be loaded during fresh startup.

Here's a part of debug output showing the proper reload sequence (I don't include debug trace facilities for now though, that is a subject for a separate PR):
```
GS [reload] begin ══════════════════════════════════════════════════════
reloading plugin GitSavvy.git_savvy
GS [reload]  ┡━─ GitSavvy.git_savvy
GS [reload]  ╿ ┡━─ GitSavvy.common
GS [reload]  ╿ ┡━─ GitSavvy.common.commands
GS [reload]  ╿ ╿ ┡━─ GitSavvy.common.commands.debug
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.common.util
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.parse_diff
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.dates
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.view
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.file
GS [reload]  ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor
GS [reload]  ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.error
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.tokens
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.events
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.nodes
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.loader
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.reader
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.scanner
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.parser
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.composer
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.constructor
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.resolver
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.dumper
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.emitter
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.serializer
GS [reload]  ╿ ╿ ╿ ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.vendor.yaml.representer
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.log
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.actions
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.debug
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.diff_string
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.util.reload
GS [reload]  ╿ ╿ ┡━─ GitSavvy.common.commands.log
GS [reload]  ╿ ╿ ┡━─ GitSavvy.common.commands.view_manipulation
GS [reload]  ╿ ╿ ┡━─ GitSavvy.common.commands.help
GS [reload]  ╿ ┡━─ GitSavvy.common.ui
GS [reload]  ╿ ┡━─ GitSavvy.common.global_events
GS [reload]  ╿ ┡━─ GitSavvy.core
GS [reload]  ╿ ┡━─ GitSavvy.core.commands
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.quick_stage
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.core.git_command
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.status
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.active_branch
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.branches
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.stash
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.stage_unstage
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.checkout_discard
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.remotes
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.ignore
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.tags
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.history
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.rewrite
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.core.git_mixins.merge
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.inline_diff
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.common.theme_generator
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.core.constants
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.commit
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.quick_commit
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.log_graph
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.checkout
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.fetch
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.pull
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.push
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.ignore
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.init
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.diff
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.blame
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.show_commit
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.branch_commit_history
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.log
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.merge
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.changelog
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.status_bar
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.reset
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.remote
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.custom
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.flow
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.cherry_pick
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.tag
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.show_file_at_commit
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.git_add
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.commands.navigate
GS [reload]  ╿ ┡━─ GitSavvy.core.interfaces
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.interfaces.status
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.interfaces.branch
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.interfaces.rebase
GS [reload]  ╿ ╿ ┡━─ GitSavvy.core.interfaces.tags
GS [reload]  ╿ ┡━─ GitSavvy.github
GS [reload]  ╿ ┡━─ GitSavvy.github.commands
GS [reload]  ╿ ╿ ┡━─ GitSavvy.github.commands.open_on_remote
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.github.github
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.common.interwebs
GS [reload]  ╿ ╿ ╿ ┡━─ GitSavvy.github.git_mixins
GS [reload]  ╿ ╿ ╿ ╿ ┡━─ GitSavvy.github.git_mixins.remotes
GS [reload]  ╿ ╿ ┡━─ GitSavvy.github.commands.commit
GS [reload]  ╿ ╿ ┡━─ GitSavvy.github.commands.configure
GS [reload]  ╿ ╿ ┡━─ GitSavvy.github.commands.add_fork_as_remote
GS [reload]  ╿ ╿ ┡━─ GitSavvy.github.commands.pull_request
GS [reload] end ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

Please find more detailed info in comments and commit log messages.

### CODE REVIEW

File: common/commands/debug.py [hunk 1]
```diff
@@ -3,45 +3,28 @@
...
-    def ordered_modules(self):
-        # common.ui must always be loaded first because it sets up the
-        # array of interface listeners.  If it's loaded after any of the
-        # interface modules then refresh events will be broken.
-        modules = ["GitSavvy.common.ui"]
-        modules += [mod for mod in sys.modules.keys()
-                    if mod[0:8] == "GitSavvy" and mod not in modules]
-
-        return modules
+    def is_visible(self):
```
Comment (stoivo) [inline common/commands/debug.py]: This is a good think.

**Comment by stoivo** [comment]
Thanks for contribution.

This is very interesting. I have tested it, a lot fast than what we had and does the save job. It looks good to me but I am not that experience with python so we will see what @divmain have to say.

**Comment by abusalimov** [comment]
@stoivo, thanks for testing it!

**Comment by divmain** [comment]
Agreed, this is a fantastic change.  Thank you!

**Pull request merged**